TodoAppState.js

```diff
@@ -0,0 +1,162 @@
+import {clearFilterForm, displayTodos, closeCreateTodoDialog, toggleFilterStatusPanel} from './view-functions.js';
+import {createTodo, readTodos, updateTodo, deleteTodo} from './mock-functions.js';
+import {Todo, todoStatuses, categories, priorities} from './todo-model-functions.js';
+
+function TodoAppState() {   
+    this.todos = [];
+    this.filters = {  
+        pattern: '',
+        todoStatus: todoStatuses.DEFAULT,
+        priorityFilter: priorities.DEFAULT,
+        categoryFilter: categories.DEFAULT
+    };
+
+    this.getTodosAndDisplay = () => {
+        readTodos()
+        .then(res => res.data)
+        .then(todosReceived => {
+            this.todos = todosReceived;
+            displayTodos([...this.todos], {...this.filters});
+        })
+        .catch(e => {
+            alert(e.message);
+        });
+    }
+
+    this.saveTodoElementHandler = () => {
+        const title = document.querySelector('#create-todo-title').value;
+        const body = document.querySelector('#create-todo-body').value;
+        const category = document.querySelector('#create-todo-category-choices').value;
+        const priority = document.querySelector('#create-todo-priority-choices').value;
+
+        if(!title) {
+            throw new Error("Empty Todo Title");
+        }
+            
+        const todo = new Todo(title, body, priority, category);
+
+        createTodo(todo)
+            .then(_ => {
+                closeCreateTodoDialog();
+            })
+            .catch(_ => {
+                alert(e.message);
+            })
+            .finally(_ => {
+                this.getTodosAndDisplay();
```

Comment:
Can we not call `this.getTodosAndDisplay` ?
`createTodo` api can return the created todo and then update the UI if successful
If `createTodo` api call fails, then handle error

Reply:
The issue with calling `this.getTodosAndDisplay();` again is that, it fetches all the `todos` and updates the whole list of UI. Fetching all todos is not required and since we are fetching all the `todos`, its payload might be more and cause unnecessary latency in real world problem. 
Updating the whole UI on single entity change is also a bad performance wise. If rendering a single todo was a costly operation, updating the full may lead to jank.
Do not update the full list if there is change in single entity